### ISSUE TITLE
Include BouncyCastle from Orbit again

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-platform/eclipse.platform.releng.aggregator#418]
See discussion on https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/403.

In short, there are two issues with Maven version of BouncyCastle:

1) The certificate that's used is not rooted in the Java 17.0.3 cacerts so installing these or updating to these will result in a prompt for for the certificate, see https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/403#issuecomment-1186481035
2) Maven version contains some parts restricted from redistribution, see https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/3008#note_916904

We should ideally fix this for 4.25 M2, latest for M3.

**Comment by iloveeclipse** [comment]
[Thanks to Matthias](https://bugs.eclipse.org/bugs/show_bug.cgi?id=580438#c8), bouncycastle 1.71.0 is now available in https://download.eclipse.org/tools/orbit/downloads/drops/I20220724195112/

**Comment by merks** [comment]
@iloveeclipse I'd really like to avoid this disruptive change appearing m2.  But the necessary changes are spread across multiple repositories and I'm not sure which order the changes needed to be done to fix this problem...

**Comment by iloveeclipse** [comment]
> I'd really like to avoid this disruptive change appearing m2

Ed, by "this change" did you mean reverting back to Orbit or the original change to Maven?

**Comment by merks** [comment]
Change to use a properly signed Orbit version, probably 1.71.0 given it's the latest available...

**Comment by mickaelistria** [comment]
I'm not convinced that reverting to Orbit's variant is a solution: same problem will happen on every bouncycastle update, and it's a pity if we keep re-packaging Orbit because of a bug in p2. Can't we fix p2 so it works before m2? Can't we add the JCE certificate to the cacerts of JustJ so it's approved by default for users of JustJ (vast majority of users)?
And after all, is it really a big issue that users are requested to trust the JCE certificate?

To be honest, I think that moving back to Orbit is the opposite of the best service for committers and users, I'm not very enthusiast about regressing here.

**Comment by merks** [comment]
Of course the Orbit version is a solution.    There is is no bug in p2 nor anywhere else in this regard.   Even the provider of this library suggests to `counter sign the BC jars` as Orbit already does.

  https://github.com/bcgit/bc-java/issues/1193#issuecomment-1190209938

I would ask you to please carefully read the following to understand that the JCE certificate is not intended to be used as a general code signing certificate:

https://www.oracle.com/java/technologies/javase/getcodesigningcertificate.html

That is why it's not in the Java cacerts and why it should not be in the Java cacerts.  I will definitely not be messing with the cacerts of a JustJ JRE which come directly from Temurin unaltered currently.

Yes, in my opinion, prompting a user to trust a certificate that is not intended to be used for that purpose is inappropriate at best and is an egregious regression.   The price of consuming these libraries directly from Maven is simply too high and is, contrary to your suggestion, is a complete disservice to the committers and the users.

**Comment by mickaelistria** [comment]
> There is is no bug in p2 nor anywhere else in this regard

The bug is that either PGP signature is not produced, or not read, as an alternative to jarsigner content. If PGP signature were used, then the jar would be automatically trusted, no pop-up, no problem.

**Comment by merks** [comment]
As I also mentioned previously, the current logic in org.eclipse.equinox.internal.p2.engine.phases.CertificateChecker.checkCertificates(SignedContentFactory) would not consider the PGP signature even if there were one present, which is not the case and not currently supported by Tycho as far as I understand.  Moreover, updating from 2022-06 to whatever we ship for 2022-09 will not have whatever fix we might add to Tycho and to p2 such that a PGP signature would be present and would actually solve the problem of prompting.   (And this all presumes that no one is going to have problems at runtime with these jars in this form.

So please do not ship something for 2022-09 that will cause problems for updating from 2022-06.   If the need to consume directly from Maven, even though EGit already does all the Orbit work for it, is so very compelling, then fix all the problems first in this release cycle and then ship such a version in the 2022-12 release cycle when the code based in prepared for it.

**Comment by mickaelistria** [comment]
I'm aware that the current logic fully ignore PGP if jarsigner is used. However, maybe it's just time to revise it to better embrace such cases.

I've parsed https://www.oracle.com/java/technologies/javase/getcodesigningcertificate.html and fail at seeing why and how it should not be allowed to add it in cacerts.

If it helps, it looks like the following command can be used to get a certificate than could then be added to cacerts if we want to trust upstream bouncycastle out of the box
```
$ unzip -p ~/apps/eclipse-SDK/plugins/bcpg_1.71.0.jar META-INF/*.DSA | openssl pkcs7 -inform DER -text -print_certs
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 4368124386800514003 (0x3c9eb1fc89f733d3)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: O=Oracle Corporation, OU=Java Software Code Signing, CN=JCE Code Signing CA
        Validity
            Not Before: Jul  6 23:48:44 2016 GMT
            Not After : Dec 31 00:00:00 2030 GMT
        Subject: O=Oracle Corporation, OU=Java Software Code Signing, CN=JCE Code Signing CA
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:87:1a:ef:2e:6c:d2:26:3f:12:d8:0c:80:4d:50:
                    52:c0:c0:53:83:34:dc:12:4e:ab:09:a9:c7:55:eb:
                    a1:95:ed:55:da:ea:78:89:49:70:28:fb:9c:fa:53:
                    2d:77:51:40:37:28:3a:99:d7:1a:b2:27:8f:64:b4:
                    02:f6:92:68:fc:88:c7:2a:a9:0a:86:c3:a1:30:57:
                    5f:cc:b2:00:31:73:53:57:50:b2:1a:4a:1a:b2:a7:
                    9c:ad:d8:b8:fa:db:42:79:14:89:4c:36:35:df:5a:
                    d2:5f:94:e5:8c:50:f9:cc:8c:4d:6c:1d:fe:b4:38:
                    62:1e:0e:2a:a0:e2:ea:52:96:22:34:54:39:da:3b:
                    9c:be:ea:e2:6b:a8:05:ee:91:04:10:8d:8c:2e:f7:
                    c6:e8:2b:b3:3c:57:6b:11:db:04:0d:87:4d:e8:20:
                    53:09:08:5f:47:9b:ae:ab:e4:9e:77:00:37:c4:65:
                    1b:8f:be:07:f1:b9:92:5a:1c:22:db:a8:e5:e9:b2:
                    9d:e9:2c:c3:d1:8e:14:6c:a1:8f:d4:0b:34:df:c8:
                    dc:00:d6:44:b2:3b:a4:78:de:b2:33:a9:1b:e2:dd:
                    e3:97:40:d2:19:c2:73:d7:cc:6f:c7:e3:c6:14:d0:
                    0d:c8:e0:a3:99:d4:da:e4:a5:ef:02:72:27:aa:c4:
                    73:01
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                59:D5:01:84:D3:02:3B:8D:41:97:46:97:E7:A5:51:07:BC:6C:20:2B
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:0
            X509v3 Authority Key Identifier: 
                59:D5:01:84:D3:02:3B:8D:41:97:46:97:E7:A5:51:07:BC:6C:20:2B
            X509v3 Key Usage: critical
                Digital Signature, Certificate Sign, CRL Sign
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        3f:13:7f:d7:c2:40:e8:1e:ce:79:86:10:6b:92:b7:fb:6b:d5:
        c0:56:0c:11:3a:84:cc:28:97:be:c8:96:a0:6c:7a:7b:b2:d2:
        62:23:25:a4:24:30:0e:86:32:cf:69:ee:50:1d:db:6f:51:00:
        d6:8d:14:89:ac:af:ff:3b:93:18:b8:e3:a1:ad:70:b5:3c:79:
        1b:ee:22:83:a5:7e:6a:19:7f:57:a6:c6:4d:3e:9d:86:80:de:
        54:c5:00:a5:b9:6d:3c:a2:7c:79:b4:98:35:1a:19:1d:b8:e4:
        21:e8:0d:d6:62:40:f6:11:3e:3a:97:ed:d8:ff:6b:69:3f:cc:
        24:9c:63:75:51:8f:fd:de:a9:2c:14:14:bb:fb:50:ef:ff:1f:
        b0:42:2d:17:34:a3:14:43:94:c2:c5:c3:f6:4c:3c:cd:35:4e:
        16:28:28:df:a6:43:b4:cb:8c:30:76:ac:f8:5b:6a:be:31:00:
        1d:43:79:5a:27:0e:d8:7c:1e:50:53:4e:ca:26:68:26:41:9d:
        c9:c5:bf:55:70:45:f7:57:b5:15:b8:31:6f:79:da:3d:b6:7f:
        39:48:af:95:44:e8:16:32:02:48:b0:e5:ab:c4:f1:11:a3:61:
        21:0e:c1:d4:9c:70:c6:75:b7:bd:6a:2b:46:61:5e:94:c8:38:
        69:03:5a:c0
-----BEGIN CERTIFICATE-----
MIIDqDCCApCgAwIBAgIIPJ6x/In3M9MwDQYJKoZIhvcNAQELBQAwYDEbMBkGA1UE
CgwST3JhY2xlIENvcnBvcmF0aW9uMSMwIQYDVQQLDBpKYXZhIFNvZnR3YXJlIENv
ZGUgU2lnbmluZzEcMBoGA1UEAwwTSkNFIENvZGUgU2lnbmluZyBDQTAeFw0xNjA3
MDYyMzQ4NDRaFw0zMDEyMzEwMDAwMDBaMGAxGzAZBgNVBAoMEk9yYWNsZSBDb3Jw
b3JhdGlvbjEjMCEGA1UECwwaSmF2YSBTb2Z0d2FyZSBDb2RlIFNpZ25pbmcxHDAa
BgNVBAMME0pDRSBDb2RlIFNpZ25pbmcgQ0EwggEiMA0GCSqGSIb3DQEBAQUAA4IB
DwAwggEKAoIBAQCHGu8ubNImPxLYDIBNUFLAwFODNNwSTqsJqcdV66GV7VXa6niJ
SXAo+5z6Uy13UUA3KDqZ1xqyJ49ktAL2kmj8iMcqqQqGw6EwV1/MsgAxc1NXULIa
Shqyp5yt2Lj620J5FIlMNjXfWtJflOWMUPnMjE1sHf60OGIeDiqg4upSliI0VDna
O5y+6uJrqAXukQQQjYwu98boK7M8V2sR2wQNh03oIFMJCF9Hm66r5J53ADfEZRuP
vgfxuZJaHCLbqOXpsp3pLMPRjhRsoY/UCzTfyNwA1kSyO6R43rIzqRvi3eOXQNIZ
wnPXzG/H48YU0A3I4KOZ1Nrkpe8CcieqxHMBAgMBAAGjZjBkMB0GA1UdDgQWBBRZ
1QGE0wI7jUGXRpfnpVEHvGwgKzASBgNVHRMBAf8ECDAGAQH/AgEAMB8GA1UdIwQY
MBaAFFnVAYTTAjuNQZdGl+elUQe8bCArMA4GA1UdDwEB/wQEAwIBhjANBgkqhkiG
9w0BAQsFAAOCAQEAPxN/18JA6B7OeYYQa5K3+2vVwFYMETqEzCiXvsiWoGx6e7LS
YiMlpCQwDoYyz2nuUB3bb1EA1o0Uiayv/zuTGLjjoa1wtTx5G+4ig6V+ahl/V6bG
TT6dhoDeVMUApbltPKJ8ebSYNRoZHbjkIegN1mJA9hE+Opft2P9raT/MJJxjdVGP
/d6pLBQUu/tQ7/8fsEItFzSjFEOUwsXD9kw8zTVOFigo36ZDtMuMMHas+FtqvjEA
HUN5WicO2HweUFNOyiZoJkGdycW/VXBF91e1Fbgxb3naPbZ/OUivlUToFjICSLDl
q8TxEaNhIQ7B1JxwxnW3vWorRmFelMg4aQNawA==
-----END CERTIFICATE-----

Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            08:87:4f:23:f4:bb:f6:3b:d8:06:a7:ae:b0:a1:2c:f4:67:2b:ba:2a
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: O=Oracle Corporation, OU=Java Software Code Signing, CN=JCE Code Signing CA
        Validity
            Not Before: Jan 25 00:58:59 2022 GMT
            Not After : Jan 25 00:58:59 2027 GMT
        Subject: O=Oracle Corporation, OU=Java Software Code Signing, CN=Legion of the Bouncy Castle Inc.
        Subject Public Key Info:
            Public Key Algorithm: dsaEncryption
                Public-Key: (2048 bit)
                pub: 
                    55:5e:7d:fb:4b:89:47:c4:13:30:81:ff:67:03:94:
                    68:86:82:c3:19:3e:6e:77:96:be:86:8e:c7:5c:b7:
                    75:95:7e:b9:ef:96:b3:cb:ee:f7:d2:9f:d4:a0:aa:
                    af:2b:ba:f8:68:53:c8:58:b3:f5:aa:59:7a:92:e7:
                    fd:c8:4a:03:84:b3:83:ec:9d:9f:93:b9:64:08:6c:
                    dd:8e:ba:f4:43:61:87:19:63:34:73:ba:a4:09:8b:
                    db:f7:b7:6c:b2:14:93:03:3f:39:41:f3:b5:35:00:
                    fa:30:b0:62:98:1f:02:57:86:b4:67:02:cf:57:00:
                    72:be:e1:7f:04:0d:23:86:27:00:0d:f0:f5:97:00:
                    30:3d:19:9c:40:fe:8f:d4:f9:38:ab:2e:81:c3:c4:
                    f2:88:eb:c5:f7:b9:c3:e3:57:13:c0:7d:6c:4b:66:
                    35:7c:93:0d:d3:58:6a:8d:d5:2a:ea:18:21:2e:e9:
                    1a:bc:6b:c3:cb:4d:91:8d:42:a3:9c:8b:a2:39:2a:
                    3d:08:c8:2d:a2:75:2d:0c:41:00:e1:d1:b1:3c:8b:
                    7a:78:3d:50:ee:e2:25:94:9b:8b:d8:92:e0:6a:fb:
                    0e:cd:72:0e:20:67:06:62:87:2b:0b:e8:7d:0d:f1:
                    5d:94:d4:15:c2:ac:da:2d:f7:13:d4:91:bb:ec:6c:
                    25
                P:   
                    00:95:47:5c:f5:d9:3e:59:6c:3f:cd:1d:90:2a:dd:
                    02:f4:27:f5:f3:c7:21:03:13:bb:45:fb:4d:5b:b2:
                    e5:fe:1c:bd:67:8c:d4:bb:dd:84:c9:83:6b:e1:f3:
                    1c:07:77:72:5a:eb:6c:2f:c3:8b:85:f4:80:76:fa:
                    76:bc:d8:14:6c:c8:9a:6f:b2:f7:06:dd:71:98:98:
                    c2:08:3d:c8:d8:96:f8:40:62:e2:c9:c9:4d:13:7b:
                    05:4a:8d:80:96:ad:b8:d5:19:52:39:8e:ec:a8:52:
                    a0:af:12:df:83:e4:75:aa:65:d4:ec:0c:38:a9:56:
                    0d:56:61:18:6f:f9:8b:9f:c9:eb:60:ee:e8:b0:30:
                    37:6b:23:6b:c7:3b:e3:ac:db:d7:4f:d6:1c:1d:24:
                    75:fa:30:77:b8:f0:80:46:78:81:ff:7e:1c:a5:6f:
                    ee:06:6d:79:50:6a:de:51:ed:bb:54:43:a5:63:92:
                    7d:bc:4b:a5:20:08:67:46:17:5c:88:85:92:5e:bc:
                    64:c6:14:79:06:77:34:96:99:0c:b7:14:ec:66:73:
                    04:e2:61:fa:ee:33:b3:cb:df:00:8e:0c:3f:a9:06:
                    50:d9:7d:39:09:c9:27:5b:f4:ac:86:ff:cb:3d:03:
                    e6:df:c8:ad:a5:93:42:42:dd:6d:3b:cc:a2:a4:06:
                    cb:0b
                Q:   
                    00:f8:18:36:68:ba:5f:c5:bb:06:b5:98:1e:6d:8b:
                    79:5d:30:b8:97:8d:43:ca:0e:c5:72:e3:7e:09:93:
                    9a:97:73
                G:   
                    42:de:bb:9d:a5:b3:d8:8c:c9:56:e0:87:87:ec:3f:
                    3a:09:bb:a5:f4:8b:88:9a:74:aa:f5:31:74:aa:0f:
                    be:7e:3c:5b:8f:cd:7a:53:be:f5:63:b0:e9:85:60:
                    32:89:60:a9:51:7f:40:14:d3:32:5f:c7:96:2b:f1:
                    e0:49:37:0d:76:d1:31:4a:76:13:7e:79:2f:3f:0d:
                    b8:59:d0:95:e4:a5:b9:32:02:4f:07:9e:cf:2e:f0:
                    9c:79:74:52:b0:77:0e:13:50:78:2e:d5:7d:df:79:
                    49:79:dc:ef:23:cb:96:f1:83:06:19:65:c4:eb:c9:
                    3c:9c:71:c5:6b:92:59:55:a7:5f:94:cc:cf:14:49:
                    ac:43:d5:86:d0:be:ee:43:25:1b:0b:22:87:34:9d:
                    68:de:0d:14:44:03:f1:3e:80:2f:41:46:d8:82:e0:
                    57:af:19:b6:f6:27:5c:66:76:c8:fa:0e:3c:a2:71:
                    3a:32:57:fd:1b:27:d0:63:9f:69:5e:34:7d:8d:1c:
                    f9:ac:81:9a:26:ca:9b:04:cb:0e:b9:b7:b0:35:98:
                    8d:15:bb:ac:65:21:2a:55:23:9c:fc:7e:58:fa:e3:
                    8d:72:50:ab:99:91:ff:bc:97:13:40:25:fe:8c:e0:
                    4c:43:99:ad:96:56:9b:e9:1a:54:6f:49:78:69:3c:
                    7a
        X509v3 extensions:
            X509v3 Authority Key Identifier: 
                59:D5:01:84:D3:02:3B:8D:41:97:46:97:E7:A5:51:07:BC:6C:20:2B
            X509v3 Extended Key Usage: 
                Code Signing
            X509v3 Subject Key Identifier: 
                2E:D1:58:A7:86:CE:6F:DF:52:78:66:1A:FE:9D:47:C8:54:83:4A:10
            X509v3 Key Usage: critical
                Digital Signature
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        28:92:06:11:63:ac:b9:78:ac:d3:44:57:4c:91:76:1c:7d:33:
        ce:49:5f:50:98:9b:6d:0e:46:26:1c:52:9a:76:dc:6b:9b:f4:
        23:4e:aa:8b:82:04:7f:b8:9f:d4:da:e9:f4:49:9c:40:f3:1e:
        52:29:28:03:0d:bf:ab:d0:06:8b:1c:3d:b5:db:21:d4:31:a9:
        40:75:cd:92:9d:9d:52:ee:46:76:23:61:7b:58:8a:57:8c:e6:
        2f:fa:15:ff:2b:2f:43:58:1a:3c:07:88:cb:51:1c:8e:1f:25:
        25:93:4a:38:63:3c:7b:f2:c8:d6:97:1c:23:bb:79:13:61:40:
        23:ec:bd:82:57:c7:b3:96:e3:c8:27:61:0d:dd:20:71:84:a8:
        dd:c9:d2:f3:d2:3b:77:64:33:d3:7b:64:3a:17:a5:02:71:82:
        52:46:ed:da:de:d2:cc:ed:93:1f:a5:fa:3d:69:0e:76:cb:b1:
        e9:b0:76:ac:8b:ed:23:a7:04:c8:be:e7:7e:c5:51:fb:f7:8c:
        6a:d8:f5:35:17:1e:fe:0d:4a:aa:08:49:67:6e:99:6e:26:e6:
        0f:ee:5f:bc:16:eb:21:36:70:72:9d:3c:62:74:d9:e6:dc:74:
        a4:a4:bd:1d:14:b1:dc:91:cf:37:eb:85:a9:d3:de:f3:f4:e9:
        cd:03:4c:ee
-----BEGIN CERTIFICATE-----
MIIF5jCCBM6gAwIBAgIUCIdPI/S79jvYBqeusKEs9GcruiowDQYJKoZIhvcNAQEL
BQAwYDEbMBkGA1UECgwST3JhY2xlIENvcnBvcmF0aW9uMSMwIQYDVQQLDBpKYXZh
IFNvZnR3YXJlIENvZGUgU2lnbmluZzEcMBoGA1UEAwwTSkNFIENvZGUgU2lnbmlu
ZyBDQTAeFw0yMjAxMjUwMDU4NTlaFw0yNzAxMjUwMDU4NTlaMG0xGzAZBgNVBAoM
Ek9yYWNsZSBDb3Jwb3JhdGlvbjEjMCEGA1UECwwaSmF2YSBTb2Z0d2FyZSBDb2Rl
IFNpZ25pbmcxKTAnBgNVBAMMIExlZ2lvbiBvZiB0aGUgQm91bmN5IENhc3RsZSBJ
bmMuMIIDRjCCAjkGByqGSM44BAEwggIsAoIBAQCVR1z12T5ZbD/NHZAq3QL0J/Xz
xyEDE7tF+01bsuX+HL1njNS73YTJg2vh8xwHd3Ja62wvw4uF9IB2+na82BRsyJpv
svcG3XGYmMIIPcjYlvhAYuLJyU0TewVKjYCWrbjVGVI5juyoUqCvEt+D5HWqZdTs
DDipVg1WYRhv+Yufyetg7uiwMDdrI2vHO+Os29dP1hwdJHX6MHe48IBGeIH/fhyl
b+4GbXlQat5R7btUQ6Vjkn28S6UgCGdGF1yIhZJevGTGFHkGdzSWmQy3FOxmcwTi
YfruM7PL3wCODD+pBlDZfTkJySdb9KyG/8s9A+bfyK2lk0JC3W07zKKkBssLAiEA
+Bg2aLpfxbsGtZgebYt5XTC4l41Dyg7FcuN+CZOal3MCggEAQt67naWz2IzJVuCH
h+w/Ogm7pfSLiJp0qvUxdKoPvn48W4/NelO+9WOw6YVgMolgqVF/QBTTMl/Hlivx
4Ek3DXbRMUp2E355Lz8NuFnQleSluTICTweezy7wnHl0UrB3DhNQeC7Vfd95SXnc
7yPLlvGDBhllxOvJPJxxxWuSWVWnX5TMzxRJrEPVhtC+7kMlGwsihzSdaN4NFEQD
8T6AL0FG2ILgV68ZtvYnXGZ2yPoOPKJxOjJX/Rsn0GOfaV40fY0c+ayBmibKmwTL
Drm3sDWYjRW7rGUhKlUjnPx+WPrjjXJQq5mR/7yXE0Al/ozgTEOZrZZWm+kaVG9J
eGk8egOCAQUAAoIBAFVefftLiUfEEzCB/2cDlGiGgsMZPm53lr6Gjsdct3WVfrnv
lrPL7vfSn9Sgqq8ruvhoU8hYs/WqWXqS5/3ISgOEs4PsnZ+TuWQIbN2OuvRDYYcZ
YzRzuqQJi9v3t2yyFJMDPzlB87U1APowsGKYHwJXhrRnAs9XAHK+4X8EDSOGJwAN
8PWXADA9GZxA/o/U+TirLoHDxPKI68X3ucPjVxPAfWxLZjV8kw3TWGqN1SrqGCEu
6Rq8a8PLTZGNQqOci6I5Kj0IyC2idS0MQQDh0bE8i3p4PVDu4iWUm4vYkuBq+w7N
cg4gZwZihysL6H0N8V2U1BXCrNot9xPUkbvsbCWjZzBlMB8GA1UdIwQYMBaAFFnV
AYTTAjuNQZdGl+elUQe8bCArMBMGA1UdJQQMMAoGCCsGAQUFBwMDMB0GA1UdDgQW
BBQu0Vinhs5v31J4Zhr+nUfIVINKEDAOBgNVHQ8BAf8EBAMCB4AwDQYJKoZIhvcN
AQELBQADggEBACiSBhFjrLl4rNNEV0yRdhx9M85JX1CYm20ORiYcUpp23Gub9CNO
qouCBH+4n9Ta6fRJnEDzHlIpKAMNv6vQBoscPbXbIdQxqUB1zZKdnVLuRnYjYXtY
ileM5i/6Ff8rL0NYGjwHiMtRHI4fJSWTSjhjPHvyyNaXHCO7eRNhQCPsvYJXx7OW
48gnYQ3dIHGEqN3J0vPSO3dkM9N7ZDoXpQJxglJG7dre0sztkx+l+j1pDnbLsemw
dqyL7SOnBMi+537FUfv3jGrY9TUXHv4NSqoISWdumW4m5g/uX7wW6yE2cHKdPGJ0
2ebcdKSkvR0UsdyRzzfrhanT3vP06c0DTO4=
-----END CERTIFICATE-----

```

**Comment by merks** [comment]
It clearly states `A certificate received from this process will not work for anything other than authenticating JCE providers to the JCE framework (that is, the certificate will not work for deployment purposes.)`   That turns out to be true.  But you'd rather ignore that, and push it into the cacerts anyway because of a failure to see that this is not the intent of that certificate; it's purely for JCE verification. Not only that, pushing it into the cacerts will only help a small portion of our consumers.  Furthermore, as I already mentioned, and I do not feel comfortable pushing cacerts into the JustJ JRE;  I'd suggest that you attempt to convince Temurin to do that for you instead and come back when you are successful.

You've been explicitly advised by the provider to counter sign the library, but you want to ignore that too.   What can we all ignore in terms of new problems here, and how best can we shift the problem away from the source to some other target? It seems every problem can just be ignored or is not a big deal.  It's almost as if avoiding the use the Orbit version is something that must be achieved at any and all costs.  Why is that?  That doesn't seem reasonable.

**Comment by mickaelistria** [comment]
I'm still trying to consider possible alternatives before reverting to Orbit.
Another one would be to enforce addition of the BouncyCastle certificate in the preferences; but it would require that this enforcement (eg some plugin Activator or startup action) is pre-installed before upgrading BouncyCastle. While it's technically doable, I'm not sure we have a good process for that. It's basically like releasing a maintenance release.

> It's almost as if avoiding the use the Orbit version is something that must be achieved at any and all costs. Why is that? 

It's indeed the goal we have in our team @RedHat (after having led Orbit for several years). The reasons why we think Orbit should be avoided at any and all costs are, for cases where the artifact pre-exists in Maven repo with OSGi information:
1. Orbit provides no functional or technical added value over upstream OSGi artifacts
2. Orbit takes much more resource when we want to update to a newer version of an artifact (basically several man.hours or even man.day while plain Maven is a matter of man.minutes)
3. Orbit adds a delay between the will for a project to consume an artifact and its capability to do so (basically takes weeks instead of minutes)
4. Orbit sometimes generates incorrect OSGi metadata compared to upstream: missing `uses` directives, missing other metadata, incorrect package versions...
5. Because it creates a different "branch" and most of Eclipse RCP community works on that branch, Orbit reduces the capability of our community to better interact with the wider world. More specifically, Orbit doesn't contribute to improving how upstream projects generate and publish OSGi metadata
6. We can expect that upstream artifacts are usually better tested than Orbit's forked ones, thus we can usually expect higher quality by bypassing Orbit

So investing in replacing Orbit by upstream Maven does immediately provide a huge return on investment; and is likely to make Eclipse RCP world closer to the "real" Java world.
Of course, Orbit was a good idea back then, before OSGi got so popular as it is now and when OSGi metadata were about never produced by upstream project, and there are also some corner cases (eg BouncyCastle?) that might still make it relevant here and there, but the big trend is that the least we rely on alternative strategy and the more we rely on upstream/standard approach, the more productive and reactive we are, and the better we integrate in the whole Java ecosystem.

**Comment by cdietrich** [comment]
i had already asked this before: why leave the "just use maven" to the individual projects and not do it on orbit.

**Comment by akurtakov** [comment]
> i had already asked this before: why leave the "just use maven" to the individual projects and not do it on orbit. the man minutes summed up to n projects also adds up.

The answer for me is:
* It doubles(could even triple) the time I have to spend to update library
* Orbit doesn't even compile using Java 17+ so I have to remember to switch jvm for it only 
* Orbit jarsigns artifacts which creates new and different artifacts from upstream ones

All of these are surely fixable but no one stepped up to do so and convince me that there is still future for Orbit.

**Comment by iloveeclipse** [comment]
Note : one point is still in progress: IP check, so actually we shouldn't move without that at very beginning.
In general, what is our strategy regarding IP checks & move to Maven based artifacts?
Did we evaluated if all the replaced libraries that we use now from Maven are actually OK from IP point of view?

In the concrete case:
 - we have someone who maintains Orbit because the artifact is used by JGit too
 - we (platform) don't need to spent any time deploy anything because it is already there
 - we need to re-sign the artifact because of certificate issue
 - this (certificate issue) will unlikely change in the future, see comments from Ed why
 - IP check question was not yet resolved
 - and we have a regression

So what kind of solution do we have, without using Orbit?

**Comment by merks** [comment]
The time being invested here in order to avoid Orbit seems to be without bound, which seems questionable at best.    It's pointless to digress into the pros and cons of Orbit, all issues which could arguably be fixed, because we'll be arguing ourselves blue in the face about that.  The point is that RedHat has made its decisions for all of us; end of discussion; don't go there.

In this case, given that the work has already been done by Mathias Sohn and was done by him in the past, there really isn't a long list of gripes to address for this specific case.   Switching to the Orbit version is quick and easy.  Not only that, you have plenty or Orbit dependencies remaining:

https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/blob/527d545ed217e6e86dea31f9f122af9732eaaf23/eclipse.platform.releng.prereqs.sdk/eclipse-sdk-prereqs.target#L5-L88

There is mention of the "real" world and its relevant to note that in the real world, everyone updating to 2022-09 will be prompted to trust a questionable certificate; one clearly described as not being appropriate for that purpose.   No solution you've suggested fixes that after the fact for the 2022-06 release currently on people's desktops.    If you want to invest more time and effort into solving the problem(s) associated with the issues around these specific artifacts, then I would suggest that is time poorly invested, but it's of course your time to invest as you see fit.   Keep in mind that this is will be a recurring problem when certificates expire and new ones are issued.

All I ask is that for the 2022-09 SimRel, please do not ship this version with this problem.  If you feel so compelled, implement solutions and ship those solutions for 2022-09.  I.e., first pave the way to the brighter future that you envision...   Please, pretty please.  If I need to debase myself and beg, let me know how best to do so...

**Comment by akurtakov** [comment]
> Note : one point is still in progress: IP check, so actually we shouldn't move without that at very beginning. In general, what is our strategy regarding IP checks & move to Maven based artifacts? Did we evaluated if all the replaced libraries that we use now from Maven are actually OK from IP point of view?

Yes, we rely on Dash License tool https://github.com/eclipse/dash-licenses/ it's integrated as a verification for releng.aggregator e.g. https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/runs/7549796075?check_suite_focus=true . Note that CQs are no longer required ahead of time and the single non-vetted dependency (osgi.annotation.bundle) is in progress https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/3082 .

**Comment by mickaelistria** [comment]
> one point is still in progress: IP check, so actually we shouldn't move without that at very beginning.

Wrong. We have IP check automation using the dash-license tool in the build, and I did run the dash-license tool manually to verify before doing the change. The dash-license tool, which is the reference and recommended tool by Eclipse Foundation staff, was pretty clear: no review was considered necessary. So the IP is clean enough and was verified and verification was completed several times already.

> In general, what is our strategy regarding IP checks & move to Maven based artifacts?

The recommendation from Eclipse Foundation is to use dash-license tool. Platform has automated of IP checks with dash-license tool on every Pull Request against the eclipse.platform.releng.aggregator Git repo several months ago.

> Did we evaluated if all the replaced libraries that we use now from Maven are actually OK from IP point of view?

I did, and all the automation we have did it too.

> So what kind of solution do we have, without using Orbit?

One more sustainable solution is to automatically trust this certificate by adding it in p2 preferences like other trusted certificates that are not in cacerts and user decides to trust. Another more sustainable solution would be to tweak p2 so it considers the PGP signature as well and add Platform's PGP signature for BouncyCastle.
But it looks like none of those solutions can currently help for 2022-09 as they need first a released fix before we can use Orbit's variant.
So the current most probable path forward, unless we get a better idea in the meantime, is to revert to Orbit variant and prepare p2 right now so it can trust based on PGP signatures even if certificates is not trusted.

**Comment by akurtakov** [comment]
All deps properly vetted now https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/runs/7556433159?check_suite_focus=true

**Comment by akurtakov** [comment]
> The point is that RedHat has made its decisions for all of us; end of discussion; don't go there.

I'm very sad to see such comments!!!
We have made this decision so we can continue to release, so we can continue working on Eclipse IDE, so we can preserve the current manpower we spend on Eclipse IDE as we have to keep delivering according to our management requirements. We have made this decision as the people throwing out the release, as the people doing the extra steps when last release broke the (Maven) world and etc. 
Surely, if enough other people step up to do the endless maintenance work, ensure builds are coming, build machines are aligned with user ones, releng scripts work and so on and on so our work in this area is no longer needed we would happily step back from these duties and go and work on more interesting and "profitable" areas!

**Comment by akurtakov** [comment]
I would really enjoy if someone else does this work and we can concentrate on the very pressing issue with SWT on GTK4, which itself will need couple of years to shape up properly so we don't end up with urged switch like it happened with GTK3.

**Comment by mickaelistria** [comment]
FWIW, replace Orbit by Central has also been desired and supported by non-Red Hat committers who are taking care of the Eclipse Platform build, and also of JDT as it will facilitate the updates of JUnit or ASM which used to be very tedious.

I mentioned our team at Red Hat because when you work as a team and get frequent team reports, then you realize more easily where the energy is spent and where the value is created, and get more insights about what to improve in the processes. But the technical reasons that apply to our team do apply to the community as a whole.

> Surely, if enough other people step up to do the endless maintenance work, ensure builds are coming, build machines are aligned with user ones, releng scripts work and so on and on so our work in this area is no longer needed we would happily step back from these duties and go and work on more interesting and "profitable" areas!

I even challenge that we would be so selfish as to leaving other committers take care of such tedious work if we can invest a bit to prevent them from having to do it while achieving higher quality by the way. Community productivity is actually something that matters as much, or even more, than our own productivity in practice.

**Comment by merks** [comment]
> > The point is that RedHat has made its decisions for all of us; end of discussion; don't go there.
> 

This is in response to `It's indeed the goal we have in our team at Red Hat..`.  I'm not sure why Red Hat needs to be mentioned.  But as I said, let's not go there. 

> I'm very sad to see such comments!!! We have made this decision so we can continue to release, so we can continue working on Eclipse IDE, so we can preserve the current manpower we spend on Eclipse IDE as we have to keep delivering according to our management requirements. We have made this decision as the people throwing out the release, as the people doing the extra steps when last release broke the (Maven) world and etc. Surely, if enough other people step up to do the endless maintenance work, ensure builds are coming, build machines are aligned with user ones, releng scripts work and so on and on so our work in this area is no longer needed we would happily step back from these duties and go and work on more interesting and "profitable" areas!

This is why I say, the decision is made, let's not go there again.  But this was presented (and continues to presented) as Red Hat's goals and opinions by someone choosing to speak on behalf of Red Hat.  I think "we" should avoid speaking opinions other than our own.  Sorry for any offense taken.

Indeed, we all all doing the best we can with the resources available to provide goodness to the community and we can **all** be commended for that effort.  Sometimes opinions on how best to do that will differ, which is fine.  As you know, I've invested a lot of effort to help ensure that the PGP support is of industrial-quality such that we are on a good path forward toward this common goal that was decided some months ago.  This annoying problem is just an unanticipated glitch and we will get past it in due course.

**Comment by merks** [comment]
> I would really enjoy if someone else does this work and we can concentrate on the very pressing issue with SWT on GTK4, which itself will need couple of years to shape up properly so we don't end up with urged switch like it happened with GTK3.

I did offer to do the work, but asked about the order of events/changes required because this spans multiple repos:

https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/418#issuecomment-1195083160

The offer still stands...

**Comment by merks** [comment]
It looks like Mikael has started the process:

https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/428

Thanks!

I can also offer to help with this because that looks to be the best solution going forward:

https://github.com/eclipse-equinox/p2/issues/121

If we had a repo with a bundle that is both PGP signed and jar signed, I can offer to do some preliminary prototype work on the problem...

**Comment by mickaelistria** [comment]
I'm indeed taking care of the revert.
I don't think I'll be able to work on https://github.com/eclipse-equinox/p2/issues/121 soon, as I will be on holiday almost for the next 3 weeks and have a few other things I want to clear before that. So if you're available for it, your help is really welcome.

**Comment by merks** [comment]
Very good!  Let's continue the discussion there then.

**Comment by akurtakov** [comment]
First of all, I'm very thankful Ed for the work you spent on PGP story!
Second, no offense taken. Actually, I'm trying to be extremely open and set clear expectations. As much as each of us has it's own opinion on different topics in the end of the day we have to do what we are paid for. 
Whenever Mickael and me speak on behalf of Red Hat it is because this direction is set for us in one way or another for the whole "Red Hat Eclipse Team" (note other teams may and probably have different agendas!) and it's better to share that publicly so others know to expect work towards same directions from other members of that team. I thought that way we make the communication easier. The goal has never been to give more weight to smth just because "Red Hat wants it that way" but rather to be "expect people from that team to be working towards that goal" so if one has other goals he/she knows to step up and do work towards his/her goals.
Time spent on releng work is supposed to reduce (and preferably significantly so) with each release, time spent on supporting the IDE to run on old versions of libraries, JVMs, other tools and etc. is supposed to be almost zero, shipped outdated libraries as part of the IDE are supposed to not happen and etc. are some of the constraints we have to work within.
